### PR TITLE
feat(ui): Require aria-label on buttons without children [INGEST-866]

### PR DESCRIPTION
**Pull request by matejminar**
This PR adds [Discriminated unions](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) to alter the Button's type to require aria-label prop if there are no meaningful children.

One example of that would be icon buttons:
```
// invalid
<Button icon={<IconDelete />} />

// valid
<Button icon={<IconDelete />} aria-label="Delete" />

// valid
<Button>Delete</Button>
```

### CODE REVIEW

File: static/app/components/button.tsx [hunk 1]
```diff
@@ -82,16 +89,11 @@ function BaseButton({
     return disabled ? undefined : prop;
   }
 
-  // For `aria-label`
-  const screenReaderLabel =
-    ariaLabel || (typeof children === 'string' ? children : undefined);
```
Comment (matejminar) [inline static/app/components/button.tsx:87]: No need to set aria-label explicitly. Scree readers natively understand this scenario.

File: static/app/components/confirm.tsx [hunk 2]
```diff
@@ -315,6 +320,7 @@ class ConfirmModal extends React.Component<ModalProps, ModalState> {
                 priority={priority}
                 onClick={this.handleConfirm}
                 autoFocus
+                aria-label={typeof confirmText === 'string' ? confirmText : t('Confirm')}
```
Comment (priscilawebdev) [inline static/app/components/confirm.tsx:323]: so it's ok to have `aria-label` and `children` with the same text?

When writing tests, should we give preference to aria-labels or texts?

File: static/app/components/button.tsx [hunk 3]
```diff
@@ -17,6 +17,16 @@ import {Theme} from 'sentry/utils/theme';
...
 type ButtonElement = HTMLButtonElement & HTMLAnchorElement & any;
 
+type ConditionalAriaLabel =
+  | {
+      children: Omit<React.ReactNode, 'null' | 'undefined' | 'boolean'>;
+      'aria-label'?: string;
+    }
+  | {
+      children?: null | boolean;
+      'aria-label': string;
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply static/app/components/button.tsx:27]: just to keep the pattern :D

File: static/app/components/confirm.tsx (hunk 2 above)
Comment (matejminar) [reply static/app/components/confirm.tsx:323]: > so it's ok to have aria-label and children with the same text?

It's not wrong, just unnecessary. I'm doing it here to assure typescript that at least some aria-label will be present.

> When writing tests, should we give preference to aria-labels or texts?

It heavily depends on the situation:
https://develop.sentry.dev/frontend/using-rtl/#querying

**Review by priscilawebdev — APPROVED** [review]
lgtm!

**Pull request merged**